prism_compile.c

```diff
@@ -729,30 +729,62 @@ pm_compile_while(rb_iseq_t *iseq, int lineno, pm_node_flags_t flags, enum pm_nod
     return;
 }
 
-static void
+static int
 pm_interpolated_node_compile(pm_node_list_t *parts, rb_iseq_t *iseq, NODE dummy_line_node, LINK_ANCHOR *const ret, const uint8_t *src, bool popped, pm_scope_node_t *scope_node, pm_parser_t *parser)
 {
+    int number_of_items_pushed = 0;
     size_t parts_size = parts->size;
 
     if (parts_size > 0) {
+        VALUE current_string = Qnil;
         for (size_t index = 0; index < parts_size; index++) {
             pm_node_t *part = parts->nodes[index];
 
             if (PM_NODE_TYPE_P(part, PM_STRING_NODE)) {
-                pm_string_node_t *string_node = (pm_string_node_t *) part;
-                ADD_INSN1(ret, &dummy_line_node, putobject, parse_string(&string_node->unescaped, parser));
+                pm_string_node_t *string_node = (pm_string_node_t *)part;
+                VALUE string_value = parse_string(&string_node->unescaped, parser);
+                if (RTEST(current_string)) {
+                    current_string = rb_str_concat(current_string, string_value);
+                }
+                else {
+                    current_string = string_value;
+                }
             }
             else {
+                if (RTEST(current_string)) {
+                    if (parser->frozen_string_literal) {
+                        ADD_INSN1(ret, &dummy_line_node, putobject, rb_str_freeze(current_string));
+                    }
+                    else {
+                        ADD_INSN1(ret, &dummy_line_node, putstring, current_string);
```

Comment:
I think we should freeze this too.  If anyone manages to get a reference to it, they could mutate an object that iseqs point to.  Basically, I think _any_ objects that iseqs directly point to should be frozen.  `putstring` will just dup the frozen string, so we'll still end up with a mutable string on the stack.